File: mirakc/agent.ts
/**
 * チューナー側のエージェント。mirakc を抱えて、denpa の代わりにチューナーを触る。
 *
 * denpa からは触れないものが3つある。
 *
 * - B-CASカード … pcscd 経由でしか読めず、その pcscd はこのコンテナにしか居ない
 * - チューナーデバイス … スキャンは mirakc を通さず recisdb で直接叩く
 * - mirakc の設定 … config.yml は起動時にしか読まれないので、書いたら再起動が要る
 *
 * そのため mirakc の親としてこれが PID 1 になり、必要なときに mirakc を止めて
 * スキャンし、設定を書き戻してから起動し直す。
 */

import {
    copyFileSync,
    existsSync,
    mkdirSync,
    readdirSync,
    readFileSync,
    renameSync,
    rmSync,
    writeFileSync,
} from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { parseConfig, replaceChannels } from './config';
import {
    CHANNEL_RANGES,
    type ChannelEntry,
    type ChannelType,
    channelsFor,
    Scanner,
    type Tuner,
} from './scan';

const PORT = Number(process.env.AGENT_PORT ?? 40773);
/*
 * 設定は /etc/mirakc に置かない。あそこには mirakc 自身の strings.yml が居て、
 * PVC を被せると隠れてしまい mirakc が起動しなくなる
 */
const CONFIG = process.env.MIRAKC_CONFIG ?? '/app-config/config.yml';
const CONFIG_TEMPLATE = '/app-config-defaults/config.yml';
const EPG_CACHE = process.env.MIRAKC_EPG_CACHE ?? '/var/lib/mirakc/epg';
/** denpa の生TSの置き場。掛かったままのTSは必ずここにある */
const RECORDED = resolve(process.env.RECORDED_DIR ?? '/denpa-recorded');
const RECISDB = process.env.RECISDB ?? 'recisdb';
const MIRAKC = process.env.MIRAKC_BIN ?? '/usr/local/bin/mirakc';
const LOG_LIMIT = 400;

function log(message: string): void {
    console.log(`[agent] ${message}`);
}

/** 子プロセスを最後まで回して、出力をまとめて受け取る */
async function run(command: string[], timeout?: number): Promise<{ code: number; output: string }> {
    try {
        const proc = Bun.spawn(command, { stdout: 'pipe', stderr: 'pipe' });
        const stop = timeout === undefined ? undefined : setTimeout(() => proc.kill(), timeout);
        const [stdout, stderr, code] = await Promise.all([
            new Response(proc.stdout).text(),
            new Response(proc.stderr).text(),
            proc.exited,
        ]);
        if (stop !== undefined) clearTimeout(stop);
        return { code, output: `${stdout}${stderr}`.trim() };
    } catch (error) {
        return { code: -1, output: String(error) };
    }
}

/** mirakc の面倒を見る。落ちたら起こし、頼まれたら止める */
class Mirakc {
    private proc: Bun.Subprocess | null = null;
    /** 設定を置き終わるまで起動させない。設定が無いと mirakc は即死する */
    private wanted = false;

    begin(): void {
        this.wanted = true;
        void this.supervise();
    }

    private async supervise(): Promise<void> {
        for (;;) {
            if (this.wanted && this.proc === null) {
                this.proc = Bun.spawn([MIRAKC], { stdout: 'inherit', stderr: 'inherit' });
                log(`mirakc を起動しました (pid ${this.proc.pid})`);
            }
            const proc = this.proc;
            if (proc === null) {
                await Bun.sleep(500);
                continue;
            }
            await proc.exited;
            if (this.proc === proc) {
                this.proc = null;
                // 止めたのが自分なら黙って待つ。落ちたのなら起こし直す
                if (this.wanted) {
                    log('mirakc が落ちました。起こし直します');
                    await Bun.sleep(1000);
                }
            }
        }
    }

    async stop(): Promise<void> {
        this.wanted = false;
        const proc = this.proc;
        if (proc !== null) {
            proc.kill();
            await proc.exited;
        }
        log('mirakc を止めました');
    }

    start(): void {
        this.wanted = true;
    }

    get alive(): boolean {
        return this.proc !== null;
    }
}

const mirakc = new Mirakc();

/**
 * カードリーダーが見えているか。
 *
 * pcscd が動いていてもリーダーを掴めていないことがある(USBが黙る)。
 * そうなると recisdb は黙って復号せずに素通しし、録画は成功したように見えて
 * 中身が全部スクランブルされたまま、という分かりにくい壊れ方をする。
 */
async function cardStatus() {
    const pcscd = (await run(['pgrep', '-x', 'pcscd'], 10_000)).code === 0;
    const scan = await run(['pcsc_scan', '-r'], 15_000);
    // 「0: Reader name」の形で並ぶ
    const readers = scan.output
        .split('\n')
        .filter((line) => /^\s*\d+:\s/.test(line))
        .map((line) => line.trim().replace(/^\d+:\s*/, ''));

    const message = !pcscd
        ? 'pcscd が動いていません'
        : readers.length > 0
          ? `カードリーダーが見えています (${readers.length} 台)`
          : 'pcscd は動いていますが、カードリーダーが見つかりません';
    return { ok: pcscd && readers.length > 0, pcscd, readers, message };
}

/** 置き場の中に収まるパスだけ受け付ける。外を読み書きさせない */
function inside(name: unknown): string | null {
    if (typeof name !== 'string' || name === '') return null;
    const full = resolve(RECORDED, name);
    return full.startsWith(`${RECORDED}/`) ? full : null;
}

/**
 * 掛かったまま録れたTSを解く。
 *
 * recisdb はカードが読めないとき「黙って素通しする」ので、終了コードだけでは
 * 成否が分からない。出来上がったものを見て判断するのは呼び出し側(denpa)。
 */
async function decode(body: { input?: unknown; output?: unknown }) {
    const source = inside(body.input);
    const target = inside(body.output);
    if (source === null || target === null) {
        return { ok: false, error: `生TSの置き場の外は解除に回せません` };
    }
    if (!existsSync(source)) {
        return {
            ok: false,
            error: `${source} が見えません。denpa と同じ置き場をこのコンテナにも見せてください`,
        };
    }

    const { code, output } = await run([RECISDB, 'decode', '-i', source, target]);
    if (code !== 0) return { ok: false, error: `recisdb が ${code} で終了しました\n${output}` };
    return { ok: true, error: '' };
}

interface ScanState {
    state: 'idle' | 'running' | 'done' | 'failed' | 'canceled';
    phase: string;
    log: string[];
    scanned: number;
    total: number;
    channels: number;
    error: string | null;
    startedAt: number | null;
    finishedAt: number | null;
}

let scan: ScanState = {
    state: 'idle',
    phase: '',
    log: [],
    scanned: 0,
    total: 0,
    channels: 0,
    error: null,
    startedAt: null,
    finishedAt: null,
};

function push(
    line?: string,
    counts: { scanned?: number; channels?: number; skipped?: number } = {},
    fields: Partial<ScanState> = {},
): void {
    scan = {
        ...scan,
        ...fields,
        log: line === undefined ? scan.log : [...scan.log, line].slice(-LOG_LIMIT),
        scanned: scan.scanned + (counts.scanned ?? 0) + (counts.skipped ?? 0),
        channels: scan.channels + (counts.channels ?? 0),
    };
}

function loadTuners(): Tuner[] {
    return parseConfig(readFileSync(CONFIG, 'utf8')).tuners ?? [];
}

/** 並べ替えの順。設定を読むときに種別ごとにまとまっているほうが分かりやすい */
const TYPE_ORDER: Record<string, number> = { GR: 0, BS: 1, CS: 2 };

/**
 * 見つけたチャンネルだけ差し替える。
 *
 * チューナーの定義はハードウェアの話で、スキャンで分かるものではない。
 * epg やサーバの設定ごと書き換えると、スキャンのたびに設定が飛ぶ。
 *
 * **探した種別だけ**を入れ替え、他はそのまま残す。地上波だけスキャンしたときに
 * 全部を置き換えると、BS と CS が設定から消える(実際に消して、BSの予約が
 * 録れなくなった)。
 */
function saveChannels(channels: ChannelEntry[], scanned: ChannelType[]): void {
    const source = readFileSync(CONFIG, 'utf8');
    const kept = (parseConfig(source).channels ?? []).filter((channel) => !scanned.includes(channel.type));
    const merged = [...kept, ...channels].sort(
        (a, b) => (TYPE_ORDER[a.type] ?? 9) - (TYPE_ORDER[b.type] ?? 9) || a.channel.localeCompare(b.channel),
    );
    const text = replaceChannels(source, merged);

    // 書きかけを読ませない。mirakc は起動時にしか読まないので、壊れたものを
    // 掴むと起動しなくなる
    const working = `${CONFIG}.writing`;
    writeFileSync(working, text);
    try {
        renameSync(working, CONFIG);
    } catch {
        // config.yml をファイル単位で bind mount していると差し替えられない
        // (compose の例がその形)。その場合は諦めて直接書く
        writeFileSync(CONFIG, text);
        rmSync(working, { force: true });
    }
}

/**
 * mirakc が覚えている局と時刻を捨てる。
 *
 * services.json などにスキャン前の局が残っていると、消えたはずの局が
 * 番組表に出続ける。次の起動で拾い直させる。
 */
function clearEpgCache(): void {
    if (!existsSync(EPG_CACHE)) return;
    for (const name of readdirSync(EPG_CACHE)) {
        rmSync(join(EPG_CACHE, name), { force: true });
    }
}

/** 走っているスキャン。中断のために持っておく */
let scanner: Scanner | null = null;

async function runScan(targets: [ChannelType, string[]][]): Promise<void> {
    const running = new Scanner(loadTuners(), (progress) => push(progress.line, progress));
    scanner = running;
    try {
        // 選局は mirakc を通さず recisdb を直接叩く。動かしたままだと EPG 更新と
        // チューナーの取り合いになるので、スキャンの間だけ止める
        push('mirakc を止めています...', {}, { phase: 'mirakc を停止' });
        await mirakc.stop();

        push('チャンネルを探しています...', {}, { phase: 'スキャン中' });
        const found = await running.run(targets);

        /*
         * 中断されたら**何も書かない**。
         * 途中までの結果で上書きすると、まだ回っていない種別やチャンネルの
         * 定義が消える。押した人は「やめたい」だけで「消したい」ではない
         */
        if (running.aborted) {
            push('中断しました', {}, { state: 'canceled', phase: '中断', finishedAt: Date.now() });
            return;
        }

        // 1件も見つからないまま上書きすると、今まで録れていた局まで消える
        if (found.length === 0) {
            throw new Error('チャンネルが1件も見つかりませんでした。チューナーとアンテナを確認してください');
        }

        push('設定を書き込んでいます...', {}, { phase: '設定を反映' });
        saveChannels(
            found,
            targets.map(([type]) => type),
        );
        clearEpgCache();
        push('mirakc を起動しています...', {}, { state: 'done', phase: '完了', finishedAt: Date.now() });
    } catch (error) {
        push(`失敗しました: ${error}`, {}, { state: 'failed', error: String(error), finishedAt: Date.now() });
    } finally {
        scanner = null;
        // 中断しても失敗しても、mirakc は必ず戻す。止まったままだと録画も番組表も死ぬ
        mirakc.start();
    }
}

function stopScan() {
    if (scan.state !== 'running' || scanner === null)
        return { stopped: false, message: '実行中ではありません' };
    scanner.stop();
    push('中断しています...', {}, { phase: '中断中' });
    return { stopped: true, message: 'チャンネルスキャンを中断しています' };
}

function startScan(body: { types?: unknown }) {
    const requested = Array.isArray(body.types) ? body.types : ['GR', 'BS', 'CS'];
    const types = requested.filter((type): type is ChannelType => type in CHANNEL_RANGES);
    if (types.length === 0) return { started: false, message: 'チャンネル種別の指定が不正です' };

    // 範囲は決め打ち。放送で使う物理チャンネルは決まっていて、狭めても
    // 総当たりの時間が少し減るだけ。狭めた結果 見つからない局が出るほうが困る
    const targets: [ChannelType, string[]][] = types.map((type) => [type, channelsFor(type)]);

    if (scan.state === 'running') return { started: false, message: '既に実行中です' };
    scan = {
        state: 'running',
        phase: '準備中',
        log: [],
        scanned: 0,
        total: targets.reduce((sum, [, channels]) => sum + channels.length, 0),
        channels: 0,
        error: null,
        startedAt: Date.now(),
        finishedAt: null,
    };
    void runScan(targets);
    return { started: true, message: 'チャンネルスキャンを始めました' };
}

const json = (body: unknown, status = 200) =>
    new Response(JSON.stringify(body), { status, headers: { 'Content-Type': 'application/json' } });

export function serve(port = PORT): Bun.Server {
    return Bun.serve({
        port,
        hostname: '0.0.0.0',
        // スキャンは何十分もかかる。既定のまま切られると解除も途中で落ちる
        idleTimeout: 0,
        async fetch(request) {
            const { pathname } = new URL(request.url);

            if (pathname === '/denpa/card' && request.method === 'GET') return json(await cardStatus());
            if (pathname === '/denpa/scan' && request.method === 'GET') {
                return json({ ...scan, mirakc: mirakc.alive });
            }
            if (pathname === '/denpa/decode' && request.method === 'POST') {
                const result = await decode(await request.json());
                return json(result, result.ok ? 200 : 500);
            }
            if (pathname === '/denpa/scan' && request.method === 'POST') {
                const result = startScan(await request.json());
                return json(result, result.started ? 200 : 409);
            }
            if (pathname === '/denpa/scan/stop' && request.method === 'POST') {
                const result = stopScan();
                return json(result, result.stopped ? 200 : 409);
            }
            return json({ ok: false, error: 'not found' }, 404);
        },
    });
}

if (import.meta.main) {
    if (!existsSync(CONFIG) && existsSync(CONFIG_TEMPLATE)) {
        mkdirSync(dirname(CONFIG), { recursive: true });
        copyFileSync(CONFIG_TEMPLATE, CONFIG);
        log(`初期設定を置きました: ${CONFIG}`);
    }
    mkdirSync(EPG_CACHE, { recursive: true });

    if ((await run(['pgrep', '-x', 'pcscd'], 10_000)).code !== 0) {
        Bun.spawn(['pcscd', '--foreground', '--disable-polkit'], { stdout: 'ignore', stderr: 'ignore' });
        log('pcscd を起動しました');
    }

    for (const signal of ['SIGTERM', 'SIGINT'] as const) {
        process.on(signal, () => {
            void mirakc.stop().then(() => process.exit(0));
        });
    }

    mirakc.begin();
    serve();
    log(`listening on :${PORT} (recorded: ${RECORDED})`);
}

File: mirakc/agent.py
#!/usr/bin/env python3
"""チューナー側のエージェント。mirakc を抱えて、denpa の代わりにチューナーを触る。

denpa からは触れないものが3つある。

* B-CASカード … pcscd 経由でしか読めず、その pcscd はこのコンテナにしか居ない
* チューナーデバイス … スキャンは mirakc を通さず recisdb で直接叩く
* mirakc の設定 … config.yml は起動時にしか読まれないので、書いたら再起動が要る

そのため mirakc の親としてこれが PID 1 になり、必要なときに mirakc を止めて
スキャンし、設定を書き戻してから起動し直す。
"""

import json
import os
import re
import shutil
import signal
import subprocess
import sys
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

import yaml

PORT = int(os.environ.get('AGENT_PORT', '40773'))
CONFIG = Path(os.environ.get('MIRAKC_CONFIG', '/etc/mirakc/config.yml'))
CONFIG_TEMPLATE = Path('/app-config-defaults/config.yml')
EPG_CACHE = Path(os.environ.get('MIRAKC_EPG_CACHE', '/var/lib/mirakc/epg'))
RECORDED_DIR = Path(os.environ.get('RECORDED_DIR', '/denpa-recorded')).resolve()
RECISDB = os.environ.get('RECISDB', 'recisdb')
SCANNER = os.environ.get('ISDB_SCANNER', 'isdb-scanner')
SCAN_OUTPUT = Path('/tmp/denpa-scan')

# ISDBScanner が選局する物理チャンネルの数。地上波は 13〜62ch を総当たりし、
# 衛星は同じネットワークの情報をまとめて取れるので1波につき1回で済む
GR_CHANNELS = 50
SATELLITE_CHANNELS = {'free': 1, 'all': 3}
# 進み具合はこの行を数えて出す。ISDBScanner は選局のたびにこれを出力する
CHANNEL_LINE = re.compile(r'Channel:\s')
LOG_LIMIT = 400


def log(message):
    print(f'[agent] {message}', flush=True)


def run(command, timeout=30):
    """子プロセスを最後まで回して、出力をまとめて受け取る"""
    try:
        done = subprocess.run(command, capture_output=True, text=True, timeout=timeout)
        return done.returncode, (done.stdout + done.stderr).strip()
    except (OSError, subprocess.SubprocessError) as error:
        return -1, str(error)


class Mirakc:
    """mirakc の面倒を見る。落ちたら起こし、頼まれたら止める"""

    def __init__(self):
        self.proc = None
        # 設定を置き終わるまで起動させない。設定が無いと mirakc は即死する
        self.wanted = False
        self.lock = threading.Lock()

    def begin(self):
        self.wanted = True
        threading.Thread(target=self._supervise, daemon=True).start()

    def _supervise(self):
        while True:
            with self.lock:
                if self.wanted and self.proc is None:
                    self.proc = subprocess.Popen(['/usr/local/bin/mirakc'])
                    log(f'mirakc を起動しました (pid {self.proc.pid})')
                proc = self.proc
            if proc is not None:
                proc.wait()
                with self.lock:
                    if self.proc is proc:
                        self.proc = None
                        # 止めたのが自分なら黙って抜ける。落ちたのなら起こし直す
                        if self.wanted:
                            log('mirakc が落ちました。起こし直します')
                            time.sleep(1)
            else:
                time.sleep(0.5)

    def stop(self):
        with self.lock:
            self.wanted = False
            proc = self.proc
        if proc is not None:
            proc.terminate()
            try:
                proc.wait(timeout=15)
            except subprocess.TimeoutExpired:
                proc.kill()
        log('mirakc を止めました')

    def start(self):
        with self.lock:
            self.wanted = True

    def alive(self):
        with self.lock:
            return self.proc is not None


mirakc = Mirakc()


def card_status():
    """カードリーダーが見えているか。

    pcscd が動いていてもリーダーを掴めていないことがある(USBが黙る)。
    そうなると recisdb は黙って復号せずに素通しし、録画は成功したように見えて
    中身が全部スクランブルされたまま、という分かりにくい壊れ方をする。
    """
    running = run(['pgrep', '-x', 'pcscd'], timeout=10)[0] == 0
    output = run(['pcsc_scan', '-r'], timeout=15)[1]
    # 「0: Reader name」の形で並ぶ
    readers = [
        line.strip().split(': ', 1)[1]
        for line in output.splitlines()
        if re.match(r'^\s*\d+:\s', line)
    ]
    if not running:
        message = 'pcscd が動いていません'
    elif readers:
        message = f'カードリーダーが見えています ({len(readers)} 台)'
    else:
        message = 'pcscd は動いていますが、カードリーダーが見つかりません'
    return {'ok': running and bool(readers), 'pcscd': running, 'readers': readers, 'message': message}


def inside_recorded(name):
    """生TSの置き場の中に収まるパスだけ受け付ける。外を読み書きさせない"""
    if not isinstance(name, str) or name == '':
        return None
    full = (RECORDED_DIR / name).resolve()
    return full if full.is_relative_to(RECORDED_DIR) and full != RECORDED_DIR else None


def decode(body):
    """掛かったまま録れたTSを解く。

    recisdb はカードが読めないとき「黙って素通しする」ので、終了コードだけでは
    成否が分からない。出来上がったものを見て判断するのは呼び出し側(denpa)。
    """
    source = inside_recorded(body.get('input'))
    target = inside_recorded(body.get('output'))
    if source is None or target is None:
        return {'ok': False, 'error': f'生TSの置き場 ({RECORDED_DIR}) の外は解除に回せません'}
    if not source.exists():
        return {
            'ok': False,
            'error': f'{source} が見えません。denpa と同じ生TSの置き場をこのコンテナにも見せてください',
        }

    code, output = run([RECISDB, 'decode', '-i', str(source), str(target)], timeout=None)
    if code != 0:
        return {'ok': False, 'error': f'recisdb が {code} で終了しました\n{output}'}
    return {'ok': True, 'error': ''}


scan = {
    'state': 'idle',
    'phase': '',
    'log': [],
    'scanned': 0,
    'total': 0,
    'channels': 0,
    'tuners': 0,
    'error': None,
    'startedAt': None,
    'finishedAt': None,
}
scan_lock = threading.Lock()


def scan_push(line, **fields):
    with scan_lock:
        if line:
            scan['log'] = (scan['log'] + [line])[-LOG_LIMIT:]
            if CHANNEL_LINE.search(line):
                scan['scanned'] += 1
        scan.update(fields)


def merge_config(scanned):
    """スキャン結果を今の設定に混ぜる。

    ISDBScanner が出すのは channels と tuners だけ。epg やサーバの設定は
    こちらのものを残さないと、スキャンのたびに設定が飛ぶ。
    """
    current = yaml.safe_load(CONFIG.read_text()) or {}
    found = yaml.safe_load(scanned.read_text()) or {}
    for key in ('channels', 'tuners'):
        if found.get(key):
            current[key] = found[key]
    # 書きかけを読ませない。mirakc は起動時にしか読まないので壊れると起動しなくなる
    working = CONFIG.with_suffix('.yml.writing')
    working.write_text(yaml.safe_dump(current, allow_unicode=True, sort_keys=False))
    working.replace(CONFIG)
    return len(current.get('channels') or []), len(current.get('tuners') or [])


def clear_epg_cache():
    """mirakc が覚えている局と時刻を捨てる。

    services.json などにスキャン前の局が残っていると、消えたはずの局が
    番組表に出続ける。次の起動で拾い直させる。
    """
    if not EPG_CACHE.is_dir():
        return
    for path in EPG_CACHE.iterdir():
        if path.is_file():
            path.unlink()


def run_scan(exclude_pay_tv):
    try:
        # ISDBScanner は mirakc を通さず直接チューナーを開く。動かしたままだと
        # EPG更新と取り合いになってスキャンが失敗するので、その間だけ止める
        scan_push('mirakc を止めています...', phase='mirakc を停止')
        mirakc.stop()

        shutil.rmtree(SCAN_OUTPUT, ignore_errors=True)
        command = [SCANNER, str(SCAN_OUTPUT)]
        if exclude_pay_tv:
            command.append('--exclude-pay-tv')

        scan_push('チャンネルを探しています...', phase='スキャン中')
        proc = subprocess.Popen(
            command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, bufsize=1
        )
        for line in proc.stdout:
            scan_push(line.rstrip())
        code = proc.wait()
        if code != 0:
            raise RuntimeError(f'isdb-scanner が {code} で終了しました')

        scanned = SCAN_OUTPUT / 'mirakc' / 'config.yml'
        if not scanned.is_file():
            raise RuntimeError('スキャン結果が出力されませんでした')

        scan_push('設定を書き込んでいます...', phase='設定を反映')
        channels, tuners = merge_config(scanned)
        # isdb-scanner はチューナーが1台も無くても 0 で終わる。1件も見つからない
        # ときに成功扱いにすると「スキャンしたのに空」が普通に見えてしまう
        if channels == 0:
            raise RuntimeError('チャンネルが1件も見つかりませんでした。チューナーとアンテナを確認してください')
        clear_epg_cache()
        scan_push(f'チャンネル {channels} 件 / チューナー {tuners} 件', channels=channels, tuners=tuners)
    except Exception as error:  # noqa: BLE001 - 何で失敗しても画面に理由を出したい
        scan_push(f'失敗しました: {error}', state='failed', error=str(error), finishedAt=time.time())
    else:
        scan_push('mirakc を起動しています...', state='done', phase='完了', finishedAt=time.time())
    finally:
        mirakc.start()


def start_scan(body):
    with scan_lock:
        if scan['state'] == 'running':
            return {'started': False, 'message': '既に実行中です'}
        exclude_pay_tv = body.get('excludePayTv') is True
        scan.update(
            state='running',
            phase='準備中',
            log=[],
            scanned=0,
            total=GR_CHANNELS + SATELLITE_CHANNELS['free' if exclude_pay_tv else 'all'],
            channels=0,
            tuners=0,
            error=None,
            startedAt=time.time(),
            finishedAt=None,
        )
    threading.Thread(target=run_scan, args=(exclude_pay_tv,), daemon=True).start()
    return {'started': True, 'message': 'チャンネルスキャンを始めました'}


class Handler(BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    def log_message(self, *args):
        pass

    def _send(self, status, body):
        payload = json.dumps(body).encode()
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def _body(self):
        length = int(self.headers.get('Content-Length') or 0)
        # パスと種別しか来ないので、それ以上溜まるのは何かがおかしい
        if length > 64 * 1024:
            raise ValueError('body too large')
        return json.loads(self.rfile.read(length) or b'{}')

    def do_GET(self):
        if self.path == '/denpa/card':
            self._send(200, card_status())
        elif self.path == '/denpa/scan':
            with scan_lock:
                self._send(200, {**scan, 'mirakc': mirakc.alive()})
        else:
            self._send(404, {'ok': False, 'error': 'not found'})

    def do_POST(self):
        try:
            body = self._body()
        except ValueError as error:
            self._send(400, {'ok': False, 'error': str(error)})
            return

        if self.path == '/denpa/decode':
            result = decode(body)
            self._send(200 if result['ok'] else 500, result)
        elif self.path == '/denpa/scan':
            result = start_scan(body)
            self._send(200 if result['started'] else 409, result)
        else:
            self._send(404, {'ok': False, 'error': 'not found'})


def main():
    if not CONFIG.exists() and CONFIG_TEMPLATE.exists():
        CONFIG.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy(CONFIG_TEMPLATE, CONFIG)
        log(f'初期設定を置きました: {CONFIG}')
    EPG_CACHE.mkdir(parents=True, exist_ok=True)

    if run(['pgrep', '-x', 'pcscd'], timeout=10)[0] != 0:
        subprocess.Popen(['pcscd', '--foreground', '--disable-polkit'])
        log('pcscd を起動しました')

    def shutdown(*_):
        mirakc.stop()
        sys.exit(0)

    signal.signal(signal.SIGTERM, shutdown)
    signal.signal(signal.SIGINT, shutdown)

    mirakc.begin()
    log(f'listening on :{PORT} (recorded: {RECORDED_DIR})')
    ThreadingHTTPServer(('0.0.0.0', PORT), Handler).serve_forever()


if __name__ == '__main__':
    main()
